cmd/plugins/juju-wait-for/application.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"time"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/apiserver/params"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/core/status"
+)
+
+func newApplicationCommand() cmd.Command {
+	cmd := &applicationCommand{}
+	cmd.newWatchAllAPIFunc = func() (WatchAllAPI, error) {
+		client, err := cmd.NewAPIClient()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return watchAllAPIShim{
+			Client: client,
+		}, nil
+	}
+	return modelcmd.Wrap(cmd)
+}
+
+const applicationCommandDoc = `
+Wait for a given application to reach a goal state.
+
+arguments:
+name
+   application name identifier
+
+options:
+-m, --model (= "")
+   juju model to operate in
+--state (= "active")
+   state of the application to wait-for
+`
+
+// applicationCommand stores image metadata in Juju environment.
+type applicationCommand struct {
+	waitForCommandBase
+
+	newWatchAllAPIFunc func() (WatchAllAPI, error)
+
+	Name    string
+	State   string
+	Timeout time.Duration
+}
+
+// Init implements Command.Init.
+func (c *applicationCommand) Init(args []string) (err error) {
+	if len(args) == 0 {
+		return errors.New("application name must be supplied when waiting for an application")
+	}
+	if len(args) != 1 {
+		return errors.New("only one application name can be supplied as an argument to this command")
+	}
+	c.Name = args[0]
+	return nil
+}
+
+// Info implements Command.Info.
+func (c *applicationCommand) Info() *cmd.Info {
+	return jujucmd.Info(&cmd.Info{
+		Name:    "application",
+		Purpose: "wait for an application to reach a goal state",
+		Doc:     applicationCommandDoc,
+	})
+}
+
+// SetFlags implements Command.SetFlags.
+func (c *applicationCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.waitForCommandBase.SetFlags(f)
+	f.StringVar(&c.State, "state", "active", "goal state of the application")
+	f.DurationVar(&c.Timeout, "timeout", time.Minute, "how long to wait, before timing out")
+}
+
+// Run implements Command.Run.
+func (c *applicationCommand) Run(ctx *cmd.Context) error {
+	client, err := c.newWatchAllAPIFunc()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	watcher, err := client.WatchAll()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		_ = watcher.Stop()
+	}()
+
+	var timedout bool
+	go func() {
+		select {
+		case <-time.After(c.Timeout):
+			timedout = true
+			watcher.Stop()
+		}
+	}()
+
+	var applicationFound bool
+	var applicationStatus string
+
+LOOP:
```

Comment:
Do you actually need these labels? Looks like the `break` and `continue` will apply to the correct scope without them...

Reply:
Nope, was from a different idea, I'll clean up